File: Sources/Handler/LRDictionary.m
//
//  LRDictionary
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

#import "LRDictionary.h"
#import "LRString.h"

@implementation NSDictionary (LRDictionary)

- (NSString *)queryString
{
	NSMutableString *queryString = nil;
	NSArray *keys = [self allKeys];

	if ([keys count] > 0) {
		for (id key in keys) {
			id value = [self objectForKey:key];
			if (nil == queryString) {
				queryString = [[NSMutableString alloc] init];
				[queryString appendFormat:@"?"];
			} else {
				[queryString appendFormat:@"&"];
			}

			if (nil != key && nil != value) {
				[queryString appendFormat:@"%@=%@", [key URLEncodedString], [value URLEncodedString]];
			} else if (nil != key) {
				[queryString appendFormat:@"%@", [key URLEncodedString]];
			}
		}
	}

	return queryString;
}

+ (NSDictionary *)dictionaryWithQueryString: (NSString *)queryString {
	if (queryString.length == 0) {
		return @{};
	}

	NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
	NSArray *pairs = [queryString componentsSeparatedByString:@"&"];

	for (NSString *pair in pairs)
	{
		NSArray *elements = [pair componentsSeparatedByString:@"="];
		if (elements.count == 2)
		{
			NSString *key = elements[0];
			NSString *value = elements[1];
			NSString *decodedKey = [key URLDecodedString];
			NSString *decodedValue = [value URLDecodedString];

			if (![key isEqualToString:decodedKey])
				key = decodedKey;

			if (![value isEqualToString:decodedValue])
				value = decodedValue;

			[dictionary setObject:value forKey:key];
		}
	}

	return [NSDictionary dictionaryWithDictionary:dictionary];
}

-(NSDictionary *) dictionaryWithLowercaseKeys {
    NSMutableDictionary         *result = [NSMutableDictionary dictionaryWithCapacity:0];
    NSString                    *key;

    for ( key in self ) {
        if([[self objectForKey:key] isKindOfClass:[NSArray class]] )
        {
            NSArray *array = [self objectForKey:key];
            [result setObject:[self lowercaseKeysAllDictionariesInArray:array] forKey:[key lowercaseString]];
        }else if ([[self objectForKey:key] isKindOfClass:[NSDictionary class]])
        {
            [result setObject:[[self objectForKey:key] dictionaryWithLowercaseKeys] forKey:[key lowercaseString]];
        }else
        {
            [result setObject:[self objectForKey:key] forKey:[key lowercaseString]];
        }
    }

    return result;
}

-(NSArray *) lowercaseKeysAllDictionariesInArray:(NSArray *) array
{
    NSMutableArray *result = [NSMutableArray arrayWithCapacity:0];

    for (id arr in array)
    {
        if([arr isKindOfClass:[NSArray class]] )
        {
            [result addObject:[self lowercaseKeysAllDictionariesInArray:arr]];

        }else if ([arr isKindOfClass:[NSDictionary class]])
        {
            NSDictionary *dict = arr;
            [result addObject:[dict dictionaryWithLowercaseKeys]];
        }else
        {
            [result addObject:arr];
        }
    }
    
    return [result copy];
}

@end

File: Sources/LRDictionary.swift
//
//  LRDictionary.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//


import Foundation

public extension Dictionary where Key == String, Value == Any {
    var queryString: String {
        var queryString = ""
        var isFirstKey = true
        
        for (key, value) in self {
            if let encodedKey = key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
               let encodedValue = "\(value)".addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
                if isFirstKey {
                    queryString.append("?")
                    isFirstKey = false
                } else {
                    queryString.append("&")
                }
                queryString.append("\(encodedKey)=\(encodedValue)")
            }
        }
        
        return queryString
    }
    
    public static func dictionaryWithQueryString(_ queryString: String) -> [String: Any] {
        if queryString.isEmpty {
            return [:]
        }
        
        var dictionary: [String: Any] = [:]
        let pairs = queryString.components(separatedBy: "&")
        
        for pair in pairs {
            let elements = pair.components(separatedBy: "=")
            if elements.count == 2 {
                let key = elements[0]
                let value = elements[1]
                if let decodedKey = key.removingPercentEncoding,
                   let decodedValue = value.removingPercentEncoding {
                    dictionary[decodedKey] = decodedValue
                }
            }
        }
        
        return dictionary
    }
    
    public func dictionaryWithLowercaseKeys() -> [String: Any] {
        var result: [String: Any] = [:]
        
        for (key, value) in self {
            if let valueArray = value as? [Any] {
                let lowercasedArray = lowercaseKeysAllDictionariesInArray(valueArray)
                result[key.lowercased()] = lowercasedArray
            } else if let valueDict = value as? [String: Any] {
                let lowercasedDict = valueDict.dictionaryWithLowercaseKeys()
                result[key.lowercased()] = lowercasedDict
            } else {
                result[key.lowercased()] = value
            }
        }
        
        return result
    }
    
    public func lowercaseKeysAllDictionariesInArray(_ array: [Any]) -> [Any] {
        var result: [Any] = []
        
        for item in array {
            if let itemArray = item as? [Any] {
                let lowercasedArray = lowercaseKeysAllDictionariesInArray(itemArray)
                result.append(lowercasedArray)
            } else if let itemDict = item as? [String: Any] {
                let lowercasedDict = itemDict.dictionaryWithLowercaseKeys()
                result.append(lowercasedDict)
            } else {
                result.append(item)
            }
        }
        
        return result
    }
}
